core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala

```diff
@@ -1431,21 +1434,28 @@ private[spark] class DAGScheduler(
                   // If the whole job has finished, remove it
                   if (job.numFinished == job.numPartitions) {
                     markStageAsFinished(resultStage)
-                    cleanupStateForJobAndIndependentStages(job)
-                    try {
-                      // killAllTaskAttempts will fail if a SchedulerBackend does not implement
-                      // killTask.
-                      logInfo(s"Job ${job.jobId} is finished. Cancelling potential speculative " +
-                        "or zombie tasks for this job")
-                      // ResultStage is only used by this job. It's safe to kill speculative or
-                      // zombie tasks in this stage.
-                      taskScheduler.killAllTaskAttempts(
-                        stageId,
-                        shouldInterruptTaskThread(job),
-                        reason = "Stage finished")
-                    } catch {
-                      case e: UnsupportedOperationException =>
-                        logWarning(s"Could not cancel tasks for stage $stageId", e)
+                    val removedStages = cleanupStateForJobAndIndependentStages(job)
+                    removedStages += resultStage
+                    val shouldInterrupt = shouldInterruptTaskThread(job)
+                    // removedStages are only used by this job. It's safe to kill speculative or
+                    // zombie tasks in these stages.
+                    logInfo(s"Job ${job.jobId} is finished. Cancelling potential speculative " +
+                      "or zombie tasks for this job")
+                    removedStages.foreach { stage =>
+                      try {
+                        // killAllTaskAttempts will fail if a SchedulerBackend does not implement
+                        // killTask.
+                        if (stage == resultStage) {
```

Comment:
seems like you should just separate this out because you are first adding it to the hashSEt and then special casing it. Just do removedStages.foreach without the resultSTage and after the loop do resultSTage